#Chapter 2 Homework
##James O. Thomas
##Computational Physics

##Exercise 2.1: Another ball dropped from a tower

A ball is again dropped from a tower of height $h$ with initial velocity zero. Write a program that asks the user to enter the height in meters of the tower and then calculates and prints the time the ball takes until it hits the ground, ignoring air resistance. use your program to calculate the time for a ball dropped from a 100 m high tower.

In [2]:
# exercise 2.1

'''
A program to calculate the time necessary for a ball dropped from a tower to hit the ground.
Inputs:
    h, the height the ball is dropped from (in meters)
Outputs:
    t, the time it takes the ball to reach the ground (in seconds)
    
James O. Thomas
Septerber 2015
'''

import numpy as np

h = float(input("Enter the height of the tower in meters: "))
g = 9.8 # m/s^2

t = (2*h/g)**0.5

print('the time it takes for the ball to reach the ground is ',t,'seconds')

Enter the height of the tower in meters: 100
the time it takes for the ball to reach the ground is  4.5175395145262565 seconds


The output from running the code above is:

Enter the height of the tower in meters: 100

the time it takes for the ball to reach the ground is  4.5175395145262565 seconds

##Exercise 2.2: Altitude of a satellite


A satellite is to be launched ito a circular orbit around the Earth so that it orbits the planet once every $T$ seconds.


a) Show that the altitude $h$ above the Earth's surface that the satellite must have is 
$$h = \left( \frac{GMT^2}{4 \pi ^2} \right) ^{1/3} - R $$
where $G = 6.67 \times 10^{-11} \rm{m} ^3 \rm{kg} ^{-1} \rm{s}^{-2} $ is Newton's gravitational constant, $M=5.97 \times 10^{24} \rm{kg}$ is the mass of the Earth, and $R = 6371 \rm{km}$ is its radius.

b) Write a program that asks the user to enter the desired value of $T$ and then calculates and prints out the correct altitude in meters.

c) Use your program to calculate the altitudes of satellites that orbit the Earth once a day (so called "geosynchronous" orbit), once every 90 minutes, and once every 45 minutes. What do you conclude from the last of these calculations?

d) Technically a geosynchronous satellite is one that orbits the Earth per _sidereal day_, which is 23.93 hours, not 24 hours. why is this? And how much difference will it make to the altitude of the satellite?

##Solution to 2.2 a)

For a satellite to remain in a stable circular orbit, the gravitational force pulling the satellite down must be equal to a centripetal force.

$$ F_{gravity} = \frac{GMm}{r^2} $$

$$F_{centripetal} = \frac{mV^2}{r}$$

setting these equal to each other and cancelling like terms, we get:

$$\frac{GM}{r} = V^2$$

For a circular orbit, the distance travelled is $2 \pi r$. This means that for a given period ($T$), the velocity the satellite travels is

$$V = \frac{\rm{distance}}{\rm{time}} = \frac{2 \pi r}{T}$$

plugging this into the above equation yields:

$$\frac{GM}{r} = \left( \frac{2 \pi r}{T} \right) ^2$$

cancellint like terms and solving for $r$ we get:

$$r^3 = \left( \frac{GMT^2}{4 \pi ^2} \right) $$

$$ r = \left( \frac{GMT^2}{4 \pi ^2} \right) ^{1/3}$$

In the above derivation, $r$ is defined as the distance from the center of the Earth to the satellite. Thus:

$$r = R + h$$

plugging this in, we get:

$$R + h = \left( \frac{GMT^2}{4 \pi ^2} \right) ^{1/3}$$

$$h = \left( \frac{GMT^2}{4 \pi ^2} \right) ^{1/3} - R $$

which is exactly what we were looking for.

In [11]:
# exercise 2.2 b)
'''
This is a short program that calculates the altitude of a satellite in a circular orbit.

Inputs:
    T, the orbital period
    units, the units associated with the input orbital period (seconds, minutes, hours, or days)

Outputs:
    T, the orbital period in seconds
    h, the altitude of the satellite necessary for the given orbital period
    
James O. Thomas
September, 2015
'''

import numpy as np

G = 6.67*10**(-11) # m^3ks^-1s^-2; Newton's gravitational constant
M = 5.97*10**(24)  # kg; mass of the Earth
R = 6371*1000      # m; radius of Earth
T = float(input("Enter the orbital period (T): "))
units = input("What Units? [s for seconds, m for minutes, h for hours, d for days] ")
if units == 'm':
    T = T*60
    print("Orbital period is ",T, " seconds")
elif units =='h':
    T = T*60*60
    print("Orbital period is ", T, " seconds")
elif units == 'd':
    T = T*60*60*24
    print("Orbital period is ",T," seconds")
elif units =='s':
    print 

h = (G*M*T**2/(4*np.pi**2))**(1./3)-R

print("The altitude of the satellite is ",h," meters")
if h < 0:
    print("This is unphysical!")

Enter the orbital period (T): 90
What Units? [s for seconds, m for minutes, h for hours, d for days]m
Orbital period is  5400.0  seconds
The altitude of the satellite is  279321.62537285965  meters


## Solution to 2.2 c)

I am not sure how to display the output when this program is run multiple times. 

* For an orbital period of 1 day, I get an altitude of 35855910 meters.

* For an orbital period of 90 minutes, I get an altitude of 279322 meters.

* For an orbital period of 45 minutes, I get an altitude of -2181560 meters. Since the altitude is negative, the satellite would have to orbit inside of the Earth, and that is not physically possible.

##Solution to 2.2 d)

A geosynchronous satellite orbits the Earth once per _sidereal day_ because the satellite is not only moving around the Earth, the Earth is also moving around the sun. More information can be found on the [Wikepedia page](https://en.wikipedia.org/wiki/Sidereal_time).

The difference is found by running the program for both times (24 hours and 23.93 hours) then finding the difference

$$\rm{difference} = h_{1\rm{ day}} - h_{1\rm{ sidereal day}}$$

$$\rm{difference} = 35855910 \rm{m} - 35773762 \rm{m} $$

$$\rm{difference} = 82418 \rm{m} \approx 82.4 \rm{km}$$

This is a rather large difference and is more than likely non negligible.

## Exercise 2.4

A spaceship travels from Earth in a straight line at relativistic speed $v$ to another planet $x$ lightyears away. Write a program to ask the user for the value of $x$ and the speed $v$ as a fraction of the speed of light $c$, then print the out the time in years that the spaceship takes to reach its destination (a) in the rest frame of the observer on Earth, and (b) as percieved by a passenger on board the ship. Use your program to calculate the ansers for a planet 10 light years away with $v = 0.99c$.

In [15]:
# solution to exercise 2.4

'''
Program to calculate the time percieved by a traveller moving at relativistic speeds

Inputs:
    x, the distance (in lightyears) from Earth to the distant planet
    v, the speed (as a fraction of c) the spaceship travels at
Outputs:
    t_Earth, the time (in years) of the flight as percieved by an observer on Earth
    t_ship, the time (in years) of the flight as percieved by a passenger on board the ship
    
James O. Thomas
September 2015
'''

import numpy as np

x = float(input("Enter the distance from Earth to the planet in light years "))
v = float(input("Enter the speed of the ship as a fraction of c (aka the beta value) the ship travels at "))

# first find the time in Earth's frame of reference
t_Earth = x/v

# now find in the moving frame of reference
t_ship = t_Earth * np.sqrt(1-v*v)

print("The time taken in the rest frame of the observer on Earth is ", t_Earth, " years.")
print("The time taken in the moving frame of reference is ", t_ship, " years.")


Enter the distance from Earth to the planet in light years 10
Enter the speed of the ship as a fraction of c (aka the beta value) the ship travels at 0.99
The time taken in the rest frame of the observer on Earth is  10.1010101010101  years.
The time taken in the moving frame of reference is  1.42492282623  years.


Running the code above generates the following output:

Enter the distance from Earth to the planet in light years 10

Enter the speed of the ship as a fraction of c (aka the beta value) the ship travels at 0.99

The time taken in the rest frame of the observer on Earth is  10.1010101010101  years.

The time taken in the moving frame of reference is  1.42492282623  years.

## Exercise 2.5: Quantum potential step

A well-known quantum mechanics problem involves a particle of mass $m$ that encounters a one-dimensional potential step, like the one shown in the book.

The particle with initial kintetic energy $E$ and wavevector $k_1 = \sqrt{2mE}/\hbar $ entersfrom the left and encounters a sudden jump on potential energy of height $V$ at a position $x=0$. By solving the Schrodinger equation, one can show that when $E>V$ the particle may either (a) pass the step, in which case it has a lower kinetic energy of $E-V$ on the other side and a correspondingly smaller wavevector of $k_2 = \sqrt{2m(E-V)}/\hbar $, or (b) it may be reflected, keeping all of its kinetic energy and an unchanged wavevector but moving in the opposite direction. The probabilities $T$ and $R$ for transmission and reflection are given by

$$T=\frac{4k_1k_2}{(k_1 + k_2)^2}$$

$$R = \left(\frac{k_1 - k_2}{k_1 + k_2} \right) ^2 $$

Suppose we have a particle with mass equal to the electron mass $ m = 9.11 \times 10 ^{-31} \rm{kg} $ and energy 10 eV encountering a potential step of height 9 eV. Write a Python program to compute and print out the transmission and reflection probabilities using the formulas above.

In [13]:
# Solution to exercise 2.5
'''
Program to calculate the transmission and reflection coefficients for a particle encountering a potential barrier.

Inputs: none
Outputs:
    T, the transmission probability
    R, the reflection probability
    
James O. Thomas
September 2015
'''

import numpy as np

mass = 9.11*10**(-31)     # kg, the mass of the particle
m = mass / 1.79e6         # convert to eV/c
E = 10.                   # eV, the energy of the particle
V = 9.                    # eV, the height of the potential step
hbarc = 0.19732697        # eV*um reduced Planck constant from Wikipedia


k_1 = np.sqrt(2*m*E)/hbarc   # wavevectors (in units of 1/micrometer)
k_2 = np.sqrt(2*m*(E-V))/hbarc

T = (4*k_1*k_2)/(k_1+k_2)**2
R = ((k_1-k_2)/(k_1+k_2))**2

print("the probability for transmission is ", T)
print("the probability for reflection is ", R)
print("the probabilty the particle does something is ", T+R, " This should equal 1")

the probability for transmission is  0.730126136388
the probability for reflection is  0.269873863612
the probabilty the particle does something is  1.0  This should equal 1


The output from running the above code:

the probability for transmission is  0.730126136388

the probability for reflection is  0.269873863612

the probabilty the particle does something is  1.0  This should equal 1

## Exercise 2.6: Planetary orbits

the orbit in space of one body around another, such as a planet around the Sun, need not be circular. In general it takes the form of an ellipse, whth the body sometimes closer in and sometimes further out. If you are given the distance $l_1$ of closest approach that a planet maeks to the Sun, also called its _perihelion_, and its linear velocity $v_1$ at perihelion, then any other property of the orbit can be calculated from these two as follows.

a) Kepler's second law tells us that the distance $l_2$ and velocity $v_2$ of the planet at its most distant point, or _aphelion_, satisfy $l_2v_2 = l_1v_1$. At the same time the total energy, kinetic plus gravitational, of the planet with velocity $v$ and distance $r$ from the Sun is given by 

$$ E = \frac{1}{2} mv^2 - G \frac{mM}{r} $$

where $m$ is the planet's mass, $M = 1.9891 \times 10^{30} \rm{kg}$ is the mass of the Sun, and $G = 6.6738 \times 10^{-11} \rm{m}^3 \rm{kg}^{-1} \rm{s}^{-2}$ is Newton's gravitational constant. Given that energy must be conserved, shwo that $v_2$ is the smaller root of the quadratic equation

$$v_2^2 - \frac{2GM}{v_1l_1}v_2 - \left[ v_1^2 - \frac{2GM}{l_1} \right] = 0 $$

Once we have $v_2$ we can calculate $l_2$ using the relation $l_2 = l_1v_1/v_2$.

b) Given the values of $v_1$, $l_1$, and $l_2$, other parameters of the orbit are given by simple formulas that can be derived from kepler's laws and the fact that the orbit is an ellipse:

* Semi-major axis: $a = \frac{1}{2}(l_1 + l_2)$,

* Semi-minor axis: $b = \sqrt{l_1l_2}$,

* Orbital period: $T = \frac{2 \pi ab}{l_1v_1} $,

* Orbital eccentricity: $ e = \frac{l_2 - l_1}{l_2 + l_1}$.

Write a program that askes the user to enter the distance to the sun and velocity at perihelion, then calculates and prints the quantities $l_1$, $v_2$, $T$, and $e$.

c) Test your program by having it calculate the properties of the orbits of the Earth (for which $l_1 = 1.4710 \times 10^{11} \rm{m}$ and $v_1 = 3.0287 \times 10^4 \rm{m/s} $) and Halley's comet ($l_1 = 8.7830 \times 10^{10} \rm{m}$ and $v_1 = 5.4529 \times 10^4 \rm{m/s} $). Among other things, you should find that the orbital period of the Earth is one year and that of hallye's comet is about 76 years.

## Solution to 2.6 a)

We start with Kepler's law and solve it for $l_2$

$$ l_2v_2 = l_1v_1 \rightarrow l_2 = l_v \frac{v_1}{v_2} \dots (1)$$

We also know from conservation of energy that

$$E_1 = E_2$$

plugging in the total energy is the kinetic plus the gravitational potential, we get

$$\frac{1}{2} m v_1^2 - G \frac{mM}{l_1} = \frac{1}{2} m v_2^2 - G \frac{mM}{l_2}$$

cancelling the $m$ and multiplying through by 2

$$v_1^2 - \frac{2GM}{l_1} = v_2^2 - \frac{2GM}{l_2}$$

pluggin in eq(1) 

$$v_1^2 - \frac{2GM}{l_1} = v_2^2 - \frac{2GM}{l_1(v_1/v_2)}$$

$$v_1^2 - \frac{2GM}{l_1} = v_2^2 - \frac{2GM}{l_1v_1}v_2$$

rearranging we get what we were looking for

$$v_2^2 - \frac{2GM}{v_1l_1}v_2 - \left[ v_1^2 - \frac{2GM}{l_1} \right] = 0$$

It can clearly be seen that one of the roots is $v_2 = v_1$. This would imply a circular orbit. Since we are studying elliptical orbits, the other root is the root we consider.

In [2]:
# Solution to 2.6 b)
'''
Program to calculate values associated with elliptical orbits

Inputs:
    l_1, the distance to the sun at perihelion
    v_1, the speed at perihelion
Outputs:
    l_2, the distance to the sun at apehelion
    v_2, the speed at apehelion
    T, the orbital period
    e, the orbital eccentricity
    
James O. Thomas
September 2015
'''
import numpy as np

# define some constants
M = 1.9891e30  # kg, the mass of the sun
G = 6.6738e-11   # m^2kg^-1s^-2, Newton's gravitational constant

# user input for values closest to the sun
l_1 = float(input("Please input the distance to the sun at perehelion in meters "))
v_1 = float(input("Please input the speed at perehelion in meters per second "))

# coefficients for quadratic equation
a = 1
b = - 2*G*M/(l_1*v_1)
c = - (v_1*v_1 - 2*G*M/l_1)
# print(a,b,c)

v_2 = (-b - np.sqrt(b*b-4*a*c))/(2*a)
# other_root = (-b + np.sqrt(b*b-4*a*c))/(2*a)   # This should be v_2 = v_1
# print(v_2, other_root)

# calculate the other values
l_2 = l_1*v_1/v_2
semi_major_axis = 0.5*(l_1 + l_2)
semi_minor_axis = np.sqrt(l_1*l_2)
T = (2*np.pi*semi_major_axis*semi_minor_axis)/(l_1*v_1)   # the orbital period
e = (l_2 - l_1)/(l_2 + l_1)

print("the distance at apehelion is ", l_2, " meters")
print("the speed at apehelion is ", v_2, " meters per second")
print("the orbital period is ", T/3600/24/365, " years")
print("the orbital eccentricity is ", e)

Please input the distance to the sun at perehelion in meters 1.4710e11
Please input the speed at perehelion in meters per second 3.0287e4
the distance at apehelion is  152027197209.0  meters
the speed at apehelion is  29305.3991773  meters per second
the orbital period is  1.00022387772  years
the orbital eccentricity is  0.0164719131347


## Solution to 2.6 c)

**Running the code for Earth yields**

Please input the distance to the sun at perehelion in meters 1.4710e11

Please input the speed at perehelion in meters per second 3.0287e4

the distance at apehelion is  152027197209.0  meters

the speed at apehelion is  29305.3991773  meters per second

the orbital period is  1.00022387772  years

the orbital eccentricity is  0.0164719131347

**Running the code for Haley's commit yields**

Please input the distance to the sun at perehelion in meters 8.7830e10

Please input the speed at perehelion in meters per second 5.4529e4

the distance at apehelion is  5.28221466088e+12  meters

the speed at apehelion is  906.680696919  meters per second

the orbital period is  76.0817006547  years

the orbital eccentricity is  0.967288912645

Both of the orbital periods match what was given in the text pretty closely

## Exercise 2.9 The Madelung constant

In condensed matter physics the Madelung constant gives the total electric potential felt by an atom in a solid. it depends on the charges on the other atoms nearby and their locations. Consider for instance solid sodium chloride -- table salt. The sodium chloride crystal has atoms arranced on a cubic lattice, bu the alternating sodium and chlorine atoms, the dodium ones having a single positive charge $+e$ and the chlorine ones a single negative charge $-e$, where $e$ is the charge on the electron. If we label each position on the lattice by three integer coordinates $(i,j,k)$, thn the sodium atoms fall at positions where $i+j+k$ is even, and the chlorin atoms at positions where $i+j+k$ is odd.

Condsider a sodium atom at the origin, $i=j=k=0$, and let us calculate the Madelung constant. if the spacing of atoms on the lattice is $a$, then the distance from the origin to the atom at position $(i,j,k)$ is 
$$\sqrt{(ia)^2 + (ja)^2 + (ka)^2} = a \sqrt{i^2 + j^2 + k^2}$$

and the potential at the orgin created by such an atom is 

$$ V(i,j,k) = \pm \frac{e}{4 \pi \epsilon _0 a \sqrt{i^2 + j^2 + k^2}}$$

with $\epsilon _0$ being the permittivity of the vacuum and the sign of the expression depending on whether $i+j+k$ is even or odd. The total potential felt by the sodium atom is then the sum of this quantity over all other atoms. Let us assume a cubic box around the sodium at the origin. with $L$ atoms in all directions. Then

$$V_{\rm{total}} = \sum^L_{i,j,k = -L; \rm{not} i=j=k=0} V(i,j,k) = \frac{e}{4 \pi \epsilon _0 a} M$$

where $M$ is the Madelung constant, at least approximately -- technically the Madelung constant is the value of $M$ when $l \rightarrow \infty $, but one can get a good approximation just by using a large value of $L$.

Write a program to calculate and pring the Madelung constant for sodium chloride use as large a value of $L$ as you can, while still having your program run in reasonable time--say in a minute or less.

In [7]:
# Solution to 2.9: The Madelung constant
'''
Program to calculate the value of the Madelung constant

Inputs: 
    L, the number of atoms in 1 direction of a cubic box
Outputs:
    M, approximation of the Madelung constant
    
Future possible improvements:
    parallelize the code to run accross multiple cores
    
James O. Thomas
September 2015
'''

import numpy as np

L = int(input("Enter the L value "))

i_values = np.arange(-L,L+1,1)
j_values = np.arange(-L,L+1,1)
k_values = np.arange(-L,L+1,1)

# print(i_values)

M = 0.0
for i in i_values:
    if i%10==0:
        print("working on row ",i," (",-L, " to ",L,")" )      # print something so I know its not frozen
    for j in j_values:
        for k in k_values:
            if (i==0 and j==0 and k==0):
                M = M           # do nothing if at the origin
            else:
                M = M + (-1)**(i+j+k)/np.sqrt((float(i))**2+(float(j))**2+(float(k))**2)

print("The approximate Madelung constant for a cube with ", L, " atoms in all directions is ", M)


Enter the L value 100
working on row  -100  ( -100  to  100 )
working on row  -90  ( -100  to  100 )
working on row  -80  ( -100  to  100 )
working on row  -70  ( -100  to  100 )
working on row  -60  ( -100  to  100 )
working on row  -50  ( -100  to  100 )
working on row  -40  ( -100  to  100 )
working on row  -30  ( -100  to  100 )
working on row  -20  ( -100  to  100 )
working on row  -10  ( -100  to  100 )
working on row  0  ( -100  to  100 )
working on row  10  ( -100  to  100 )
working on row  20  ( -100  to  100 )
working on row  30  ( -100  to  100 )
working on row  40  ( -100  to  100 )
working on row  50  ( -100  to  100 )
working on row  60  ( -100  to  100 )
working on row  70  ( -100  to  100 )
working on row  80  ( -100  to  100 )
working on row  90  ( -100  to  100 )
working on row  100  ( -100  to  100 )
The approximate Madelung constant for a cube with  100  atoms in all directions is  -1.74181981584


Running the code for an L value of 100 gives an approximate Madelung constant of -1.742 which is relatively close to the value of -1.748 found on [Wikipedia](https://en.wikipedia.org/wiki/Madelung_constant).

# Exercise 2.10: The semi-emperical mass formula

In nuclear physics, the semi-emperical mass formula is a formula for calculating the approximate nuclear binding energy $B$ of an atomic nucleus with atomic number $Z$ and mass number $A$:

$$B = a_1 A - a_2 A^{2/3} - a_3 \frac{Z^2}{A^{1/3}} - a_4 \frac{(A-2Z)^2}{A} + \frac{a_5}{A^{1/2}} $$

where, in units of millions of electron volts, the constants are $a_1 = 15.67$, $a_2 = 17.23$, $a_3 = 0.75$, $a_4 = 93.2$, and $a_5 = 0$ if $A$ is odd, $a_5 = 12.0$ if $A$ and $Z$ are both even, $a_5=-12.0$ if $A$ is even and $Z$ is odd.

**I think that the symmetry ($a_4$) and the pairing ($a_5$) terms have incorrect values. A more realistic approach would have a much smaller symmertry term (around $23$ MeV) and slightly smaller pairing term.

a) Write a program that takes as its input the values of $A$ and $Z$, and prints out the binding energy for the corresponding atom. use your program to find the binding energy of an atom with $A = 58$ and $Z = 28$. (Hint: the correct answer is around 490 MeV.)

b) Modify your program to print out not the total binding energy $B$, but the binding energy per nucleon, whih is $B/A$.

c) Now modify your program so that it takes as input just a single value of the atomic number $Z$ and then goes through all the values of $A$ from $A=Z$ to $A=3Z$, to find the one that has the largest binding energy per nucleon. This is the most stable nucleus with the given atomic number. Have your program pring out the vale of $A$ for this most stable nucleus and the value of the binding energy per nucleon.

d) Modify your program again so that, instead of taking $Z$ as the input it runs through all values of $Z$ from 1 to 100 and prings out the most stable value of $A$ for each one. At what value of $Z$ does the maximum binding energy per nucleon occur? (The true answer, in real life, is $Z=28$, which is nickel. You should find that the semi-emperical mass formula gets the answer roughly right, but not exactly.)

In [20]:
# Solution to 2.10 a)
'''
Program to calculate the total binding energy of a nucleus
Input:
    A, the mass number
    Z, the atomic number
Output:
    B, the total binding energy
'''

# get the A and Z values
A = int(input("What is the mass number (A)? "))
Z = int(input("What is the atomic number (Z)? "))

# Define the coefficients
a_1 = 15.76    # MeV, the volume term
a_2 = 17.23    # MeV, the surface term
a_3 = 0.75     # MeV, the coulomb term
a_4 = 93.2     # MeV, the symmetry term
if A%2==1:               # if A is odd
    a_5 = 0.0           
elif A%2==0 and Z%2==0:  # if A and Z are even
    a_5 = 12.0
else:                   # if A is even and Z is odd
    a_5 = -12.0
    
B = a_1*A - a_2*A**(2.0/3) - a_3*Z*Z*A**(-1./3) - a_4*(A-2*Z)**2/A + a_5*A**(-0.5)

print("The Binding energy for A = ", A, " and Z = ", Z)
print("is ", B, " MeV")


What is the mass number (A)? 58
What is the atomic number (Z)? 28
The Binding energy for A =  58  and Z =  28
is  499.1556068013682  MeV


Running the code above for A = 58 and Z = 28 gives

What is the mass number (A)? 58

What is the atomic number (Z)? 28

The Binding energy for A =  58  and Z =  28

is  499.1556068013682  MeV

In [21]:
# Solution to 2.10 b)
'''
a slight modification to make the program output the binding energy per nucleon
'''

# copied from 2.10 a)

# get the A and Z values
A = int(input("What is the mass number (A)? "))
Z = int(input("What is the atomic number (Z)? "))

# Define the coefficients
a_1 = 15.76    # MeV, the volume term
a_2 = 17.23    # MeV, the surface term
a_3 = 0.75     # MeV, the coulomb term
a_4 = 93.2     # MeV, the symmetry term
if A%2==1:               # if A is odd
    a_5 = 0.0           
elif A%2==0 and Z%2==0:  # if A and Z are even
    a_5 = 12.0
else:                   # if A is even and Z is odd
    a_5 = -12.0
    
B = a_1*A - a_2*A**(2.0/3) - a_3*Z*Z*A**(-1./3) - a_4*(A-2*Z)**2/A + a_5*A**(-0.5)

# slight modification in print statements
print("The Binding energy per nucleon for A = ", A, " and Z = ", Z)
print("is ", B/A, " MeV")


What is the mass number (A)? 58
What is the atomic number (Z)? 28
The Binding energy per nucleon for A =  58  and Z =  28
is  8.606131151747729  MeV


Running the code for A = 58 and Z = 28 gives

What is the mass number (A)? 58

What is the atomic number (Z)? 28

The Binding energy per nucleon for A =  58  and Z =  28

is  8.606131151747729  MeV

In [12]:
# Solution to 2.10 c)
'''
Program to calculate the most stable nucleus for a given Z number
as calculated by the semi-emperical mass formula

Inputs:
    Z, the atomic number
Outputs:
    A, the mass number for the most stable nucleus
    BperA, the binding energy per nucleon for the most stable nucleus
    
James O. Thomas
September 2015
'''

import numpy as np

def Binding_Energy(A,Z):
    # Define the coefficients
    a_1 = 15.76    # MeV, the volume term
    a_2 = 17.23    # MeV, the surface term
    a_3 = 0.75     # MeV, the coulomb term
    a_4 = 93.2     # MeV, the symmetry term
    if A%2==1:               # if A is odd
        a_5 = 0.0           
    elif A%2==0 and Z%2==0:  # if A and Z are even
        a_5 = 12.0
    else:                   # if A is even and Z is odd
        a_5 = -12.0
    
    B = a_1*A - a_2*A**(2.0/3) - a_3*Z*Z*A**(-1./3) - a_4*(A-2*Z)**2/A + a_5*A**(-0.5)
    BperA = B/A
    return BperA

Z = int(input("Please enter the atomic number (Z) "))

A_array = np.arange(Z,3*Z+1,1)

B_value = 0
A_value = 0

for A in A_array:
    B_test = Binding_Energy(A,Z)
    if B_test > B_value:
        B_value = B_test
        A_value = A
        
print("The most stable nucleus with a charge number of ",Z)
print("has a mass number (A) of ", A_value)
print("and a binding energy per nucleon of ", B_value, " MeV")

Please enter the atomic number (Z) 28
The most stable nucleus with a charge number of  28
has a mass number (A) of  58
and a binding energy per nucleon of  8.60613115175  MeV


In [13]:
# Solution to 2.10 d)
'''
Program calculates the most stable nucleus with 1<=Z<=100 and Z<=A<=3Z

Input:
    none
Output:
    A value for the most stable configuration at each Z value as well as the Z value with the overall
    highest binding energy per nucleon
    
James O. Thomas
September 2015
'''

import numpy as np

def Binding_Energy(A,Z):
    # Define the coefficients
    a_1 = 15.76    # MeV, the volume term
    a_2 = 17.23    # MeV, the surface term
    a_3 = 0.75     # MeV, the coulomb term
    a_4 = 93.2     # MeV, the symmetry term
    if A%2==1:               # if A is odd
        a_5 = 0.0           
    elif A%2==0 and Z%2==0:  # if A and Z are even
        a_5 = 12.0
    else:                   # if A is even and Z is odd
        a_5 = -12.0
    
    B = a_1*A - a_2*A**(2.0/3) - a_3*Z*Z*A**(-1./3) - a_4*(A-2*Z)**2/A + a_5*A**(-0.5)
    BperA = B/A
    return BperA

Z_values = np.arange(1,101,1) # the Z values
Z_most_stable = 0
B_most_stable = 0
A_most_stable = 0

for Z in Z_values:
    B = 0
    A_stable = 0
    for A in np.arange(Z,3*Z+1,1):
        B_test = Binding_Energy(A,Z)
        if B_test > B:
            B = B_test
            A_stable = A
    print("The most stable nucleus with Z number of ",Z," has an A value of ",A_stable)
    if B > B_most_stable:
        Z_most_stable = Z
        A_most_stable = A_stable
        B_most_stable = B
print("The Z number with the overall lowest binding energy pur nucleon is ",Z_most_stable)
        
        

The most stable nucleus with Z number of  1  has an A value of  0
The most stable nucleus with Z number of  2  has an A value of  4
The most stable nucleus with Z number of  3  has an A value of  6
The most stable nucleus with Z number of  4  has an A value of  8
The most stable nucleus with Z number of  5  has an A value of  10
The most stable nucleus with Z number of  6  has an A value of  12
The most stable nucleus with Z number of  7  has an A value of  15
The most stable nucleus with Z number of  8  has an A value of  16
The most stable nucleus with Z number of  9  has an A value of  19
The most stable nucleus with Z number of  10  has an A value of  20
The most stable nucleus with Z number of  11  has an A value of  23
The most stable nucleus with Z number of  12  has an A value of  24
The most stable nucleus with Z number of  13  has an A value of  27
The most stable nucleus with Z number of  14  has an A value of  28
The most stable nucleus with Z number of  15  has an A value 

The above code gives the Z number where the maximum binding energy occurs as 24, which is close to the value of 28.

## Exercise 2.11: Binomial coefficients

The binomial coefficient $\left( ^n _k \right) $ is an integer equal to

$$\left(^n _k \right) = \frac{n!}{k!(n-k)!} = \frac{n \times (n-1) \times (n-2) \times \dots \times (n-k+1)}{1 \times 2 \times \dots \times k} $$

when $k \ge 1$, or $ \left( ^n _0 \right) = 1$ when $k=0$.

a) using this form for the binomial coefficient, write a Python user-defined function binomial(n,k) that calculates the binomial coefficient for given $n$ and $k$. Make sure your function returns the answer in the form of an integer (not a float) and gives the correct value of 1 for the case where $k=0$.

b) Using your function write a program to pring out the first 20 lines of "Pascal's triangle." the $n$th line of Pascal's triangle contains $n+1$ numbers, which are the coefficients $\left( ^n _0 \right) $, $\left( ^n_1 \right) $, and so on up to $\left( ^n_n \right) $. Thus the first few lines are

1 1

1 2 1

1 3 3 1

1 4 6 4 1

c) the probability that an unbiased coin, tossed $n$ times, will come up heads $k$ times is $\left( ^n _k \right) /2^n $. Write a program to calculate (a) the total probability that a coin tossed 100 times comes up heads exactly 60 times, and (b) the probability that it comes up heads 60 or more times.

In [1]:
'''
Exercise 2.11 a)
user defined function to calculate binomial coefficients

James O. Thomas
September 2015
'''

import numpy as np

def binomial(n,k):
    if k == 0:
        return 1
    else:
        numerator = 1
        denominator = 1
        # the odd stopping points for the iterators are because np.arange does not include the stop
        for i in np.arange(n,n-k,-1): 
            numerator = numerator*i
        for i in np.arange(1,k+1,1):
            denominator = denominator*i
        coefficient = int(numerator/denominator)
        return coefficient
 
x = binomial(4,4)
print(x, type(x))

1 <class 'int'>


The code above returns the correct values and data types for the first 4 rows of pascals triangle

In [54]:
'''
Exercise 2.11 b)
user defined function to calculate binomial coefficients
This program will calculate and print the first 20 rows of Pascal's triangle

the first 4 rows of the output should be

[1 1]
[1 2 1]
[1 3 3 1]
[1 4 6 4 1]

James O. Thomas
September 2015
'''
# this part is copy and pasted from part a
import numpy as np

def binomial(n,k):
    if k == 0:
        return 1
    else:
        numerator = 1
        denominator = 1
        # the odd stopping points for the iterators are because np.arange does not include the stop
        for i in np.arange(n,n-k,-1): 
            numerator = numerator*i
        for i in np.arange(1,k+1,1):
            denominator = denominator*i
        coefficient = int(numerator/denominator)
        return coefficient

# this is what I added for part b)
for i in np.arange(1,21,1):
    js = np.zeros(i+1, dtype=int)
    for j in np.arange(0,i+1):
        js[j] = binomial(i,j)
    print(js)
        



[1 1]
[1 2 1]
[1 3 3 1]
[1 4 6 4 1]
[ 1  5 10 10  5  1]
[ 1  6 15 20 15  6  1]
[ 1  7 21 35 35 21  7  1]
[ 1  8 28 56 70 56 28  8  1]
[  1   9  36  84 126 126  84  36   9   1]
[  1  10  45 120 210 252 210 120  45  10   1]
[  1  11  55 165 330 462 462 330 165  55  11   1]
[  1  12  66 220 495 792 924 792 495 220  66  12   1]
[   1   13   78  286  715 1287 1716 1716 1287  715  286  -29    4    1]
[   1   14   91  364 1001 2002 3003 3432 3003 2002 -182   41    1    0    1]
[   1   15  105  455 1365 3003 5005 6435 6435 5005 -547  -33   -2    0    1
    1]
[    1    16   120   560  1820  4368  8008 11440 12870  -395  -277   -43
     0     0     0     0     1]
[    1    17   136   680  2380  6188 12376 19448 24310   638   510     2
     1     0     1     0     0     1]
[    1    18   153   816  3060  8568 18564 31824 43758  1276   -34   -25
     3     0     0     0     0     3     1]
[     1     19    171    969   3876  11628  27132  50388 -30940  -2308
     58     48     -3      0     -1   

/home/james/anaconda3/lib/python3.4/site-packages/IPython/kernel/__main__.py:27: RuntimeWarning: overflow encountered in long_scalars
/home/james/anaconda3/lib/python3.4/site-packages/IPython/kernel/__main__.py:29: RuntimeWarning: overflow encountered in long_scalars


The above code generates the proper values for the first 20 rows of Pascal's triangle. However, there are runtime warnings indicate that I am reaching the limit of what the INT32 data type can accommodate.

In [3]:
# 2.11 c) part (a)
'''
This program will calculate the probability that an unbiased coin tossed 100 times will land up exactly 60 times

James O. Thomas
September 2015
'''

# this part is copied from part a)

import numpy as np

def binomial(n,k):
    if k == 0:
        return 1
    else:
        numerator = 1.0       # I had do drop the integer demands because this problem has numbers
        denominator = 1.0     # larger than what INT32 can handle
        # the odd stopping points for the iterators are because np.arange does not include the stop
        for i in np.arange(n,n-k,-1): 
            numerator = numerator*i
        for i in np.arange(1,k+1,1):
            denominator = denominator*i
        coefficient = (numerator/denominator)
        return coefficient
    
# This is what was added for part c) (a)

n = 100 # the total number of coin tosses
k = 60  # the number of times the coin is to land heads up

probability = binomial(n,k)/2**n   # the function from the problem statement; 
                                     # equavalent to the number of microstates in the macrostate divided by the
                                     # total number of microstates in all macrostates
        
print(probability)

0.010843866711637985


the probability is about 0.010843 (or 1.08%) that a coin tossed 100 times comes up heads exacly 60 times.

In [12]:
# 2.11 c) part (a)
'''
This program will calculate the probability that an unbiased coin tossed 100 times will land up exactly 60 times

James O. Thomas
September 2015
'''

# this part is copied from part a)

import numpy as np

def binomial(n,k):
    if k == 0:
        return 1
    else:
        numerator = 1.0       # I had do drop the integer demands because this problem has numbers
        denominator = 1.0     # larger than what INT32 can handle
        # the odd stopping points for the iterators are because np.arange does not include the stop
        for i in np.arange(n,n-k,-1): 
            numerator = numerator*i
        for i in np.arange(1,k+1,1):
            denominator = denominator*i
        coefficient = (numerator/denominator)
        return coefficient
    
# This is what was added for part c) (b)

n = 100 # the total number of coin tosses
k = np.arange(60,101,1)

micro_state_sum = 0   
                                     
for i in k:
    micro_state_sum = micro_state_sum + binomial(n,i)
    
probability = micro_state_sum/2**n

print(probability)

0.028443966820490385


When running the above code, the calculated probability is about 0.02844 (2.844%) that the coin comes up heads 60 or more times.

##Exercise 2.12: Prime numbers

The program in Example 2.8 is not a very efficient way of calculating prime numbers: it checks each number to see if it is divisible by an number less than it. we can develop a much faster program for prime numbers by making use of the following observations:

a) A number $n$ is prime if it has no prime factors less than $n$. Hence we only need to check if it is divisible by other primes.

b) If a number $n$ is non-prime, having a factor $r$, then $n = rs$, where $s$ is also a factor. If $r \ge \sqrt{n} $ then $n = rs \ge \sqrt{n}s$, which implies that $s \le \sqrt{n}$. In other words, any non-prime must have factors, and hence also prime factors, less than or equal to $\sqrt{n}$. Thus to determine if a number is prime we have to check its prime factors only up to and including $\sqrt{n}$ -- if there are none then the number is prime.

c) If we find even a single prime factor less than $\sqrt{n} $ then we know that the number is non-prime, and hence there is no need to check any further -- we can abandon this number and move on to something else.

Write a Python program that finds all the primes up to ten thousand. Create a list to store the primes, which starts out with just one prime number 2 in it. Then for each number $n$ from 3 to 10000 check whether the number is divisible by any of the primes in the list up to and including $\sqrt{n}$. As soon as you find a single prime factor you can stop checking the rest of them -- you know $n$ is not a prime. If you find no prime factors $\sqrt{n}$ or less then $n$ is prime and you should add it to the list. You can print out the list all in one go at the end of the program, or you can print out the individual number as you find them.

In [15]:
# Solution to 2.12
'''
Program to find the prime numbers between 2 and 10000
Inputs: 
    none
Outputs:
    the prime numbers up to 10000

James O. Thomas
September 2015
'''

import numpy as np

primes = [2] # the list to hold the primes

for n in np.arange(3,10001,1):          # loop over values to check
    n_max = np.sqrt(n)+1                # max value necessary to check (+1 because I am not sure about rounding)
    for p in primes:
        if p <= n_max and n%p==0:       # if p is less than or equal to sqrt(n) check the value, otherwise move on
            break                       # if n is evenly divisibly by p then n is not prime so move on
    else:                               # the location of what comes after the "break" seems off, but it works.
        primes.append(n)
        
print(primes)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997, 1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093, 1097, 1103, 1109, 1117, 1123, 1129, 1151, 1153, 1163, 1171, 1181, 1187, 1193, 1201, 1213, 1217, 12

The output from the above program is the list of the prime number through 10,000. The hardest part of this problem was figuring out which logic operator (and, or, etc) to use.